core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,579 @@
+                return;
+            }
+            Scheduler scheduler;
+            synchronized (this) {
+                scheduler = this.scheduler;
+                this.scheduler = null;
+            }
+            if (state == QueryState.FINISHED) {
+                if (scheduler != null) {
+                    scheduler.cancel();
+                }
+                stageManager.finish();
+            }
+            else if (state == QueryState.FAILED) {
+                if (scheduler != null) {
+                    scheduler.abort();
+                }
+                stageManager.abort();
+            }
+            queryStateMachine.updateQueryInfo(Optional.ofNullable(getStageInfo()));
+        });
+
+        scheduler = createScheduler();
+        queryExecutor.submit(scheduler::schedule);
+    }
+
+    private Scheduler createScheduler()
+    {
+        taskDescriptorStorage.initialize(queryStateMachine.getQueryId());
+        queryStateMachine.addStateChangeListener(state -> {
+            if (state.isDone()) {
+                taskDescriptorStorage.destroy(queryStateMachine.getQueryId());
+            }
+        });
+
+        Session session = queryStateMachine.getSession();
+        int partitionCount = getFaultTolerantExecutionPartitionCount(session);
+        Function<PartitioningHandle, BucketToPartition> bucketToPartitionCache = createBucketToPartitionCache(nodePartitioningManager, session, partitionCount);
+
+        ImmutableList.Builder<FaultTolerantStageScheduler> schedulers = ImmutableList.builder();
+        Map<PlanFragmentId, Exchange> exchanges = new HashMap<>();
+        NodeAllocator nodeAllocator = nodeAllocatorService.getNodeAllocator(session);
+
+        try {
+            // root to children order
+            List<SqlStage> stagesInTopologicalOrder = stageManager.getStagesInTopologicalOrder();
+            // children to root order
+            List<SqlStage> stagesInReverseTopologicalOrder = reverse(stagesInTopologicalOrder);
+
+            checkArgument(taskRetryAttemptsOverall >= 0, "taskRetryAttemptsOverall must be greater than or equal to 0: %s", taskRetryAttemptsOverall);
+            AtomicInteger remainingTaskRetryAttemptsOverall = new AtomicInteger(taskRetryAttemptsOverall);
+            List<Exchange> outputStages = new ArrayList<>();
+            for (SqlStage stage : stagesInReverseTopologicalOrder) {
+                PlanFragment fragment = stage.getFragment();
+
+                boolean outputStage = stageManager.getOutputStage().getStageId().equals(stage.getStageId());
+                ExchangeContext exchangeContext = new ExchangeContext(session.getQueryId(), new ExchangeId("external-exchange-" + stage.getStageId().getId()));
+                Exchange exchange = exchangeManager.createExchange(
+                        exchangeContext,
+                        partitionCount,
+                        // order of output records for coordinator consumed stages must be preserved as the stage
```

Comment:
Currently it is only implicitly required by the exchange between the last stage and a coordinator (as the last stage may produce sorted results because of ORDER BY)